File: 0133-clone-graph/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    public Node cloneGraph(Node originalNode) {
        if(originalNode == null) {
            return null;
        }
        
        Queue<Node> bfsQueue = new ArrayDeque<>(Arrays.asList(originalNode));
        Map<Node, Node> originalToCloneMap = new HashMap<>();
        originalToCloneMap.put(originalNode, new Node(originalNode.val));
        
        while(!bfsQueue.isEmpty()){
            Node currentNode = bfsQueue.poll();
            Node clonedNode = originalToCloneMap.get(currentNode);
            
            for(Node originalNeighbor : currentNode.neighbors) {
                if(!originalToCloneMap.containsKey(originalNeighbor)) {
                    Node clonedNeighbor = new Node(originalNeighbor.val);
                    originalToCloneMap.put(originalNeighbor, clonedNeighbor);
                    bfsQueue.offer(originalNeighbor);
                }
                clonedNode.neighbors.add(originalToCloneMap.get(originalNeighbor));
            }
        }
        
        return originalToCloneMap.get(originalNode);
    }
}

File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional


class Solution:
    def cloneGraph(self, node: Optional["Node"]) -> Optional["Node"]:
        if not node:
            return None

        visited = {}

        queue = deque([node])

        cloned_node = Node(node.val)
        visited[node] = cloned_node

        while queue:
            cn = queue.popleft()

            for neighbor in cn.neighbors:
                if neighbor not in visited:
                    cngbr = Node(neighbor.val)
                    visited[neighbor] = cngbr
                    queue.append(neighbor)

                visited[cn].neighbors.append(visited[neighbor])

        return visited[node]
